onnxscript/rewriter/llama_rule_sets.py

```diff
@@ -2,13 +2,84 @@
 # Licensed under the MIT License.
 from __future__ import annotations
 
+import numpy as np
+import onnx.numpy_helper
+
 import onnxscript.ir as ir
 import onnxscript.rewriter.no_op as no_op
 import onnxscript.rewriter.pattern as orp
 
 op = orp.onnxop
 
 
+class CastIdentity(orp.RewriteRuleAsClass):
+    """Replaces ``Cast(., to=to)`` by ``Identity`` if possible."""
+
+    @classmethod
+    def pattern(cls, op, x, to):
+        return op.Cast(x, to=to)
+
+    @classmethod
+    def rewrite(cls, op, x: ir.Value, to: ir.AttrInt64):
+        return op.Identity(x)
+
+    @classmethod
+    def check(cls, context, x, to) -> bool:
+        return x.dtype == to.value
+
+
+class CastCast(orp.RewriteRuleAsClass):
+    """Replaces ``Cast(Cast(X, ...), to=to)`` by ``Cast(X, to=to)``."""
+
+    @classmethod
+    def pattern(cls, op, x, to: int, to_ignored: int):
+        return op.Cast(op.Cast(x, to=to_ignored), to=to)
+
+    @classmethod
+    def rewrite(cls, op, x: ir.Value, to: ir.AttrInt64, to_ignored: ir.AttrInt64):
+        return op.Cast(x, to=to)
+
+
+class ExpandIdentity(orp.RewriteRuleAsClass):
+    """Replaces ``Expand(., shape)`` by ``Identity`` if possible."""
+
+    @classmethod
+    def pattern(cls, op, x, shape):
+        return op.Expand(x, shape)
+
+    @classmethod
+    def rewrite(cls, op, x: ir.Value, shape: ir.Value):
+        return op.Identity(x)
+
+    @classmethod
+    def check(cls, context, x, shape) -> bool:
+        if shape.const_value is None:
+            # Shape is not a constant and cannot be guessed.
+            return False
+        shape_x = x.shape
+        return shape_x.dims == tuple(shape.const_value.numpy().tolist())
+
+
+class ReshapeReshape(orp.RewriteRuleAsClass):
+    """Replaces ``Reshape(Reshape(X, ...), shape)`` by ``Reshape(X, shape)``."""
```

Comment:
I would explain handling of `-1`, `0` shapes as part of the docstring.

Reply:
done